### ISSUE TITLE
`/proc/cpuinfo` has wrong cpu frequencies

### ISSUE THREAD

**Issue by RagnarGrootKoerkamp** [htop-dev/htop#953]
On my Intel Nuc (`CPU: 11th Gen Intel(R) Core(TM) i5-1135G7 @ 2.40GHz`), `/proc/cpuinfo` always shows `2400.000` MHz for most of the cores:
```
cpu MHz         : 1100.063
cpu MHz         : 1148.182
cpu MHz         : 2400.000
cpu MHz         : 2400.000
cpu MHz         : 2400.000
cpu MHz         : 2400.000
cpu MHz         : 2400.000
cpu MHz         : 2400.000
```

`/sys/devices/system/cpu/cpu*/cpufreq/scaling_cur_freq` gives accurate values instead:
```
cat /sys/devices/system/cpu*/cpufreq/scaling_cur_freq
1156558
1101168
1108234
1200081
1200379
1200496
1201186
1201895
```

It turns out that on newer intel processors, `/proc/cpuinfo` does not report actual cpu frequencies anymore, reading [this](https://bugzilla.redhat.com/show_bug.cgi?id=1421876#c17).

The [PR implementing this](https://github.com/hishamhm/htop/pull/932) notes that `scaling_cur_freq` can be slow, but at least on this CPU, that seems to be the only place to get the correct value.

**Comment by fasterit** [comment]
This seems like a bug that should go upstream to the Linux kernel. Because if your sample reading of `/proc/cpuinfo` is what you always get, it has the correct CPU frequency for one core and it wrong for the three others.
NB: `constant_tsc` as per the linked Red Hat bug report is not a good indicator. It is present on a 7-gen Core as well and there reading the CPU frequency with our logic works fine.

**Comment by RagnarGrootKoerkamp** [comment]
Yeah maybe it should. But I don't know anything about TSC. Maybe that _is_ lower for some cores some of the time (unlikely though)? There is also the `nonstop_tsc` feature which is set for my CPU.

My laptop (`i7-10750H`) has the same behaviour:
<details>
<pre>
cpu MHz		: 2600.000
cpu MHz		: 2600.000
cpu MHz		: 2600.000
cpu MHz		: 2600.000
cpu MHz		: 2600.000
cpu MHz		: 2600.000
cpu MHz		: 2600.000
cpu MHz		: 893.481
cpu MHz		: 885.600
cpu MHz		: 891.814
cpu MHz		: 2600.000
cpu MHz		: 2600.000
</pre>
</details>

To clarify: each time I do `cat /proc/cpuinfo`, a few (1 to 3) random/different cpus show an actual frequency (which is some number close to 900M in this case), while all other cpus show the exact value of their base frequency.

**Comment by gary094** [comment]
This exact behaviour happens on my i5 1135G7 as well. However it wasn't always the case. I believe after some update a few months ago (arch linux) I noticed these wrong frequencies as well. I tried to find the changes here on `htop` but couldn't find any indication of this being changed. I gave up then

**Comment by serfreeman1337** [comment]
With KDE I can get correct frequency reporting in htop by adding widget:
- Add Widgets -> Total Cpu Use -> Configure Total Cpu Use -> Sensors Details -> Text-Only Sensors -> [Group] CPU -> [Group] Clock frequency -> OK

Then htop process needs to be restarted.
After that I can remove widget and will still get correct frequency reporitng on htop, although `/proc/cpuinfo` stays same all time.

Would like to know what's triggering htop to show actual frequency after adding monitor widget.

**Comment by bmwiedemann** [comment]
From reading https://www.reddit.com/r/openSUSE/comments/wtu2ft/cpu_frequency_is_not_minimal_in_idle_state/ my guess is that /proc/cpuinfo MHz values are only updated when a process is scheduled on a core

The best way to solve this would be to read `/sys/devices/system/cpu/cpu*/cpufreq/scaling_cur_freq` instead. It is even easier to parse - just 1 integer.

**Comment by BenBE** [comment]
> The best way to solve this would be to read `/sys/devices/system/cpu/cpu*/cpufreq/scaling_cur_freq` instead. It is even easier to parse - just 1 integer.

It's just one integer, but several files to `open`/`read`/`close`, which causes the number of syscalls to increase. This might not look like an issue at first glance, but as each context switch to kernel space is kinda costly, it's usually cheaper to read one file and parse (even slightly more complicated data) than to open several files across various directories (which you might even have to locate first).

**Comment by CodeMonkey80s** [comment]
For my `13th Gen Intel(R) Core(TM) i7-13700K` there is no such file like this one:
```
/sys/devices/system/cpu/cpu0/cpufreq/cpuinfo_cur_freq
cat: /sys/devices/system/cpu/cpu0/cpufreq/cpuinfo_cur_freq: No such file or directory
```

But current frequency seems to be stored in:

```
cat /sys/devices/system/cpu/cpu0/cpufreq/scaling_cur_freq
3400000
```

**Comment by trr** [comment]
Can't tell you why it's different on my system, but I had to find the above in:

    cat /sys/devices/system/cpu/cpufreq/policy*/scaling_cur_freq

Have a Ivy Bridge 4 core 8 thread processor and the frequency of the 8th "cpu" is shown as wrong most of the time.  In scaling_cur_freq it appears to be a real figure.

I'd concur with above notes that it may be triggered when nothing is being scheduled on that CPU at the time, and that it just instead pretends it's running at some default multiplier (in my case, the maximum turbo frequency)

**Comment by graysky2** [comment]
A me too comment, bug affects Intel N100/150:
`htop` shows the base frequency even under load:
```

    0[|||||||||||||||||||||||||||||||||100.0%||700MHz|44C] Hostname: r2max
    1[|||||||||||||||||||||||||||||||||100.0%||700MHz|44C] Tasks: 70, 45 thr, 131 kthr; 4 running
    2[|||||||||||||||||||||||||||||||||100.0%||700MHz|44C] Load average: 4.26 2.91 1.60 
    3[|||||||||||||||||||||||||||||||||100.0%||700MHz|44C] Uptime: 00:15:06
  Mem[||||||||||||||||||||||||||||||||||||||||6.89G/15.4G] PSI some CPU:     0.61%  0.56%  0.36% 
                                                           PSI some IO:      0.00%  0.00%  0.26% 
```

But `turbostat` shows the the real clockrate:
```
# turbostat --quiet --Summary --interval 6 --show Busy%,Bzy_MHz,PkgWatt,PkgTmp
Busy%	Bzy_MHz	PkgTmp	PkgWatt
99.76	2900	44	11.32
99.77	2900	46	11.79
```

Would be good to fix this bug.

Is the issue that turbostat reading `/dev/cpu/*/msr` only does so if run as root...

**Comment by trr** [comment]
In many of these this may not be a bug but a difference in methods of sampling core clock speeds.

They change rapidly and if sampled instantaneously they tend to over report because the act of waking up to sample sensors triggers activity even if it's only that speed for milliseconds.

They can be sampled as an average since last reading for more accuracy but the question arises of what to do when the cores are sleeping and effectively a zero clock rate. Include that time and you can get very low average clock speeds for mostly idle cores, but if you don't include that time you get a representative clock speed for the time that core was awake even if it was only away less than 1% of the time.
